app/Utils/SubmissionUtils.php

```diff
@@ -30,56 +34,111 @@ public static function get_xml_type(mixed $filehandle): array
             } elseif (str_contains($content, '<CoverageLog')) {
                 // Should be before coverage
                 $handler = \CoverageLogHandler::class;
                 $file = 'CoverageLog';
+                $schema_file = "{$schemas_dir}/{$file}.xsd";
             } elseif (str_contains($content, '<Coverage')) {
                 $handler = \CoverageHandler::class;
                 $file = 'Coverage';
+                $schema_file = "{$schemas_dir}/{$file}.xsd";
             } elseif (str_contains($content, '<report')) {
                 $handler = \CoverageJUnitHandler::class;
                 $file = 'CoverageJUnit';
             } elseif (str_contains($content, '<Notes')) {
                 $handler = \NoteHandler::class;
                 $file = 'Notes';
+                $schema_file = "{$schemas_dir}/{$file}.xsd";
             } elseif (str_contains($content, '<DynamicAnalysis')) {
                 $handler = \DynamicAnalysisHandler::class;
                 $file = 'DynamicAnalysis';
+                $schema_file = "{$schemas_dir}/{$file}.xsd";
             } elseif (str_contains($content, '<Project')) {
                 $handler = \ProjectHandler::class;
                 $file = 'Project';
+                $schema_file = "{$schemas_dir}/{$file}.xsd";
             } elseif (str_contains($content, '<Upload')) {
                 $handler = \UploadHandler::class;
                 $file = 'Upload';
+                $schema_file = "{$schemas_dir}/{$file}.xsd";
             } elseif (str_contains($content, '<testsuite')) {
                 $handler = \TestingJUnitHandler::class;
                 $file = 'TestJUnit';
             } elseif (str_contains($content, '<Done')) {
                 $handler = \DoneHandler::class;
                 $file = 'Done';
+                $schema_file = "{$schemas_dir}/{$file}.xsd";
             }
         }
 
         // restore the file descriptor to beginning of file
         rewind($filehandle);
 
+        // perform minimal error checking as a sanity check
+        if ($file === '') {
+            throw new CDashXMLValidationException(["ERROR: Could not determine submission"
+                                                  ." file type for: '{$xml_file}'"]);
+        }
+        if (isset($schema_file) && !file_exists($schema_file)) {
+            throw new CDashXMLValidationException(["ERROR: Could not find schema file '{$schema_file}'"
+                                                  ." to validate input file: '{$xml_file}'"]);
+        }
+
         return [
             'file_handle' => $filehandle,
             'xml_handler' => $handler,
             'xml_type' => $file,
+            'xml_schema' => $schema_file,
         ];
     }
 
+    /**
+     * Validate the given XML file based on its type
+     * @throws CDashXMLValidationException
+     */
+    public static function validate_xml(string $xml_file, string $schema_file): void
```

Comment:
What do you think about moving this to `app/cdash/xml_handlers/abstract_xml_handler.php`?  That way, the `$schema_file` argument could be stored as a class member and users of this method won't need to know about the exact details of how the validation actually occurs.

Reply:
@williamjallen, this one isn't so clear.  Do you think this whole function belongs in `abstract_...`  so that each handler would be able to check? Or is each handler just keeping it's schema file as a mostly hard-coded attribute.  That way, we'd still keep this in SubmissionUtils to find out what handler type to use and then let the class inform the schema file when it needs to.

Reply:
I think the entire function should be moved to `abstract_xml_handler.php` for the moment.  That way, it can access a protected static member containing the schema path for each handler, and the entire public interface is simply `*Handler::validate($filename)`.